call_option.go

```diff
@@ -1,118 +1,130 @@
+	// Invoke never calls ShouldRetry with nil error.
+	Retry(err error) (pause time.Duration, shouldRetry bool)
 }
 
-// Per-call configurable settings for retrying upon transient failure.
-type RetrySettings struct {
-	RetryCodes      map[codes.Code]bool
-	BackoffSettings BackoffSettings
-}
+type retryerOption func() Retryer
 
-// Parameters to the exponential backoff algorithm for retrying.
-type BackoffSettings struct {
-	DelayTimeoutSettings MultipliableDuration
-	RPCTimeoutSettings   MultipliableDuration
+func (o retryerOption) Resolve(s *CallSettings) {
+	s.Retry = o
 }
 
-type MultipliableDuration struct {
-	Initial    time.Duration
-	Max        time.Duration
-	Multiplier float64
+// WithRetry sets CallSettings.Retry to fn.
+func WithRetry(fn func() Retryer) CallOption {
+	return retryerOption(fn)
 }
 
-func (w CallSettings) Resolve(s *CallSettings) {
-	s.Timeout = w.Timeout
-	s.RetrySettings = w.RetrySettings
-
-	s.RetrySettings.RetryCodes = make(map[codes.Code]bool, len(w.RetrySettings.RetryCodes))
-	for key, value := range w.RetrySettings.RetryCodes {
-		s.RetrySettings.RetryCodes[key] = value
+// OnCodes returns a Retryer that retries if and only if
+// the previous attempt returns a GRPC error whose error code is stored in codes.
+// Pause times between retries are specified by bo.
+//
+// bo is only used for its parameters; each Retryer has its own copy.
+// codes must not be modified.
+func OnCodes(codes []codes.Code, bo Backoff) Retryer {
+	return &boRetryer{
+		backoff: bo,
+		codes:   codes,
 	}
 }
 
-type withTimeout time.Duration
-
-func (w withTimeout) Resolve(s *CallSettings) {
-	s.Timeout = time.Duration(w)
-}
-
-// WithTimeout sets the client-side timeout for API calls if the call isn't
-// retrying.
-func WithTimeout(timeout time.Duration) CallOption {
-	return withTimeout(timeout)
+type boRetryer struct {
+	backoff Backoff
+	codes   []codes.Code
 }
 
-type withRetryCodes []codes.Code
-
-func (w withRetryCodes) Resolve(s *CallSettings) {
-	s.RetrySettings.RetryCodes = make(map[codes.Code]bool)
-	for _, code := range []codes.Code(w) {
-		s.RetrySettings.RetryCodes[code] = true
+func (r *boRetryer) Retry(err error) (time.Duration, bool) {
+	c := grpc.Code(err)
+	for _, rc := range r.codes {
+		if c == rc {
+			return r.backoff.Pause(), true
+		}
 	}
+	return 0, false
 }
 
-// WithRetryCodes sets a list of Google API canonical error codes upon which a
-// retry should be attempted. If nil, the call will not retry.
-func WithRetryCodes(retryCodes []codes.Code) CallOption {
-	return withRetryCodes(retryCodes)
-}
-
-type withDelayTimeoutSettings MultipliableDuration
+// Backoff implements exponential backoff.
+// The wait time between retries is a random value between 0 and the "retry envelope".
+// The envelope starts at Initial and increases by the factor of Multiplier every retry,
+// but is capped at Max.
+type Backoff struct {
+	// Initial is the initial value of the retry envelope, defaults to 1 second.
```

Comment:
Could you clarify your third-to-last bullet point, how adding the internal `envelope` field could lead to confusion? Is your concern that `OnCodes` could be called to create a copy of `Backoff` that already has non-initial internal state? If that's the concern, how does that last bullet point address it, since there's still internal state there, but it's just public? (Is the idea that folks would explicitly set it?) I think I'm missing something....

Another idea is to have the user-configurable `Initial`, an internal state `envelope`, and then a
`(b Backoff) Copy() *Backoff` method that creates a copy with `envelope` initialized to `Initial`, and we use that in `OnCodes()`, while clarifying elsewhere that `Backoff` is stateful and most copy operations probably intend the semantics of `Copy()`